src/disp_s1/solid_earth_tides.py

```diff
@@ -0,0 +1,247 @@
+    except Exception as e:
+        raise RuntimeError(f"Error loading raster file {filename}: {e}") from e
+
+
+def transform_bounds_to_epsg4326(
+    bounds: BoundingBox, crs: rasterio.crs.CRS
+) -> BoundingBox:
+    """Transform bounds to EPSG:4326 if CRS is not geographic."""
+    if not crs.is_geographic:
+        transformed_bounds = transform_bounds(crs, "EPSG:4326", *bounds)
+        return BoundingBox(*transformed_bounds)
+    return bounds
+
+
+def generate_atr(
+    bounds: BoundingBox, height: int, width: int, margin_degrees: float = 0.1
+) -> dict:
+    """Generate the ATR dictionary for pysolid grid."""
+    min_grid_size = 25
+    height_small = max(min_grid_size, height // 100)
+    width_small = max(min_grid_size, width // 100)
+    return {
+        "LENGTH": height_small,
+        "WIDTH": width_small,
+        "X_FIRST": bounds.left - margin_degrees,
+        "Y_FIRST": bounds.top + margin_degrees,
+        "X_STEP": (bounds.right - bounds.left + 2 * margin_degrees) / width_small,
+        "Y_STEP": -(bounds.top - bounds.bottom + 2 * margin_degrees) / height_small,
+    }
+
+
+def interpolate_set_components(
+    n: np.ndarray,
+    e: np.ndarray,
+    y_arr: np.ndarray,
+    x_arr: np.ndarray,
+    set_east: np.ndarray,
+    set_north: np.ndarray,
+    set_up: np.ndarray,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Interpolate SET components to the unwrapped interferogram grid."""
+    points = np.stack((n, e), axis=-1)
+    return (
+        RegularGridInterpolator((y_arr, x_arr), set_east)(points),
+        RegularGridInterpolator((y_arr, x_arr), set_north)(points),
+        RegularGridInterpolator((y_arr, x_arr), set_up)(points),
+    )
+
+
+def run_solid_grid(az_time, lat, lon, res_lat, res_lon):
+    """Run solid_grid from pysolid."""
+    az_time = pd.to_datetime(az_time)
+    return solid_grid(
+        az_time.year,
+        az_time.month,
+        az_time.day,
+        az_time.hour,
+        az_time.minute,
+        az_time.second,
+        lat,
+        res_lat,
+        1,
+        lon,
+        res_lon,
+        1,
+    )
+
+
+def calculate_time_ranges(
+    reference_start_time: pd.Timestamp,
+    reference_stop_time: pd.Timestamp,
+    secondary_start_time: pd.Timestamp,
+    secondary_stop_time: pd.Timestamp,
+    atr: dict,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Generate time ranges for reference and secondary files."""
+    ref_time_range = pd.to_datetime(
+        np.linspace(
+            reference_start_time.value, reference_stop_time.value, atr["LENGTH"]
+        )
+    )
+    sec_time_range = pd.to_datetime(
+        np.linspace(
+            secondary_start_time.value, secondary_stop_time.value, atr["LENGTH"]
+        )
+    )
+    ref_time_tile = np.tile(ref_time_range, (atr["WIDTH"], 1)).T
+    sec_time_tile = np.tile(sec_time_range, (atr["WIDTH"], 1)).T
+    return ref_time_tile, sec_time_tile
+
+
+def calculate_solid_earth_tides_correction(
+    ifgram_filename: Filename,
```

Comment:
we might consider naming this [`like_filename` as `dolphin` does](https://github.com/search?q=repo%3Aisce-framework%2Fdolphin+like_filename&type=code) , or just [`like` as `snaphu-py` or `rasterio` use](https://github.com/isce-framework/snaphu-py/blob/b21dd6897a80f27bd75c88dce32a34d3d41bcb3a/src/snaphu/io/_raster.py#L144-L147), since it doesn't actually matter what the contents are: we are only using the size and geographic metadata from it